End to end

sorru94 opened the pull request.

harlem88 commented on end_to_end_tests/app/main/idf_component.yml:
```diff
@@ -0,0 +1,12 @@
+  protocol_examples_common:
+    path: '$IDF_PATH/examples/common_components/protocol_examples_common/'
+  astarte-platform/astarte-device-sdk-esp32:
+    version: '1.3.3'
```
Since we using a path to find the library, Is it valid to specify the version?

sorru94 replied:
Correct, it can be removed.

harlem88 commented on end_to_end_tests/app/main/include/utils.h:
```diff
@@ -0,0 +1,29 @@
+/*
+ * (C) Copyright 2024, SECO Mind Srl
```
```suggestion
 * (C) Copyright 2025, SECO Mind Srl
```

harlem88 commented on private/device_connection.h:
```diff
@@ -61,6 +61,17 @@ void device_connection_on_connected_handler(
+ */
+void device_connection_on_pubblish_handler(
+    astarte_device_handle_t device, esp_mqtt_event_handle_t mqtt_event);
+
```
typo pubblish -> publish
```suggestion
/**
 * @brief Handler for a PUBLISH event.
 *
 * @details This function can be used as a PUBLISH event handler for the Astarte MQTT client.
 *
 * @param[in] device Astarte device context.
 * @param[in] mqtt_event MQTT event from the MQTT driver.
 */
void device_connection_on_publish_handler(
    astarte_device_handle_t device, esp_mqtt_event_handle_t mqtt_event);

```

harlem88 commented on .github/workflows/end_to_end_tests.yaml:
```diff
@@ -0,0 +1,107 @@
+    concurrency:
+      group: end-to-end-test-${{ github.ref }}
+      cancel-in-progress: true
+    # container: espressif/idf:release-v5.4
```
Remove this, if it is no longer used
```suggestion
    # container: espressif/idf:release-v5.4
```

harlem88 commented on .github/workflows/end_to_end_tests.yaml:
```diff
@@ -0,0 +1,107 @@
+        path: astarte-device-sdk-esp32
+    - name: Install ESP IDF
+      run: |
+        sudo apt-get install git wget flex bison gperf python3 python3-pip python3-venv cmake ninja-build ccache libffi-dev libssl-dev dfu-util libusb-1.0-0
```
```suggestion
        sudo apt-get update
        sudo apt-get install git wget flex bison gperf python3 python3-pip python3-venv cmake ninja-build ccache libffi-dev libssl-dev dfu-util libusb-1.0-0
```

sorru94 committed:
End to end tests implementation

This is a initial implementation, only individual
datastreams have been tested.
Tests are performed in both directions from the
device to Astarte and from Astarte to the device.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

lucaato commented on end_to_end_tests/app/main/src/astarte_task.c:
```diff
@@ -0,0 +1,515 @@
+    device_config.property_unset_cbk = unset_property_callback;
+    device_config.cbk_user_data = NULL;
+    device_config.interfaces = interfaces;
+    device_config.interfaces_size = ARRAY_SIZE(interfaces);
```
nit: use directly the inline struct inizialization

lucaato commented on end_to_end_tests/app/main/src/astarte_task.c:
```diff
@@ -0,0 +1,515 @@
+        vTaskDelayUntil(&last_wake_time, CONFIG_DEVICE_POLL_PERIOD_MS / portTICK_PERIOD_MS);
+    }
+
+    ESP_LOGI(TAG, "Destroying the device.");
+    ares = astarte_device_destroy(device, 2000);
+    if (ares != ASTARTE_RESULT_OK) {
+        ESP_LOGE(TAG, "Failed in device destruction, err: %s", astarte_result_to_name(ares));
+        goto exit;
+    }
```
Does this code run ? Am I missing something or there are only two paths:
- one where the loop continues
- one where we go straight to the exit label

sorru94 replied:
No, this is dead code. Just leftover stuff from the samples.

lucaato commented on end_to_end_tests/app/main/src/astarte_task.c:
```diff
@@ -0,0 +1,515 @@
+            goto exit;
+        }
+    } else {
+        ESP_LOGI(TAG, "Sending object datastream to Astarte.");
```
We should add a todo here as a reminder

lucaato commented on end_to_end_tests/app/main/src/utils.c:
```diff
@@ -0,0 +1,209 @@
+ * @param num_bytes The resulting size of the returned array, only valid if the return is non NULL.
+ * @return The dynamically allocated and filled bytes array, or NULL upon error.
+ */
+uint8_t *utils_hexstr_to_bytes(const char *hexstr, size_t *num_bytes)
```
Why did you decide to use hex formatted strings as an interchange format ? It is cumbersome to deserialize from an hex string and implement our own function for that purpose. Esp should already have a module for base64 encoding/decoding that should work for this.

sorru94 replied:
No particular reason. It seems an easy way of passing data.
It's human readable and easy to copy paste to some online bson decoders ;)

lucaato replied:
I'd probably prefer using base64 and printing out the hex bytes for debugging purposes but since it is and e2e it's ok either way

lucaato commented on end_to_end_tests/end_to_end.py:
```diff
@@ -0,0 +1,525 @@
+    tmp_dir = init_tmp_directory()
+    realm, device_id, api_hostname, appengine_token = load_configuration()
+
+    connectivity_test = TestCase(
```
Since these testcases are quite lengthy I would split them in separate files.

lucaato commented on end_to_end_tests/end_to_end.py:
```diff
@@ -0,0 +1,525 @@
+        [
+            TestActionConnect(timeout=30),
+            TestActionSleep(seconds=2),
+            TestActionCheckDeviceStatus(
```
Would it make sense to test for the device connection in the other testcases too ?

sorru94 replied:
This test case is specifically made to check if the connection callback works properly. 
It would be out of scope of other tests to check the device status from the Astarte APIs.

harlem88 commented on .github/workflows/end_to_end_tests.yaml:
```diff
@@ -0,0 +1,107 @@
+      run: |
+        sudo apt-get update
+        sudo apt-get install git wget flex bison gperf python3 python3-pip python3-venv cmake ninja-build ccache libffi-dev libssl-dev dfu-util libusb-1.0-0
+        mkdir -p esp
+        cd esp
+        git clone -b release/v5.4 --recursive https://github.com/espressif/esp-idf.git
+        mkdir esp-tools
+        export IDF_TOOLS_PATH="$PWD/esp-tools"
+        cd esp-idf
+        ./install.sh all
+        echo "IDF_TOOLS_PATH=$IDF_TOOLS_PATH" >> "$GITHUB_ENV"
+        echo "IDF_PATH=$PWD" >> "$GITHUB_ENV"
```
Why don't we use the idf container? For example `espressif/idf:release-v5.4`, so we can skip these installation step?

sorru94 replied:
Because the container is lacking some dependencies that are required by the end to end tests.

harlem88 commented on end_to_end_tests/app/main/src/utils.c:
```diff
@@ -0,0 +1,209 @@
+/*
+ * (C) Copyright 2024, SECO Mind Srl
```
```suggestion
 * (C) Copyright 2024-2025, SECO Mind Srl
```

sorru94 committed:
Apply suggestions from code review

Signed-off-by: Simone Orru <simone.orru@secomind.com>

harlem88 commented on end_to_end_tests/app/main/src/astarte_task.c:
```diff
@@ -0,0 +1,510 @@
+#endif
+
+    // You shouldn't log a credential secret in a production device
+    ESP_LOGI(TAG, "Credential secret: '%s'", cred_secr);
```
nit: So I think you can remove it ;-)
```suggestion
```

harlem88 reviewed (changes requested):
LGTM

lucaato commented on end_to_end_tests/cases/case_reception.py:
```diff
@@ -0,0 +1,203 @@
+            datetime(2025, 4, 26, 13, 32, 0, tzinfo=timezone.utc),
+        ),
+        TestActionSleep(seconds=2),
+        TestActionReadReceivedMQTTData(
+            "org.astarteplatform.end-to-end.ServerDatastream",
+            "/binaryblob_endpoint",
+            b"binblob",
+        ),
```
I have a strong opinion on these steps... probably because I'm biased by how I handled this in the other e2e tests, but splitting the same message in two test actions causes a lot of repetition. I don't think it's handy, especially if you need to edit the test cases, so transmit and receive should be merged in a single step. Since I believe this is quite the shift, this probably can't be done right now, but we should keep it in mind going forward.

sorru94 replied:
I think the end-to-end tests in Zephyr took a very different approach.
The focus of the end-to-end tests in this repository is to provide the user with the most granularity possible when setting up a test case.
Keeping the transmit and check reception separated allows the user to have more control over when they want to perform the check. This could be useful in more complex test cases that hopefully will be added in the future.

lucaato commented on end_to_end_tests/app/main/src/utils.c:
```diff
@@ -0,0 +1,209 @@
+            }
+            break;
+        case ASTARTE_MAPPING_TYPE_DATETIME:
+            int64_t datetime = false;
```
Initialize to `0` instead of `false`

lucaato commented on end_to_end_tests/app/main/src/utils.c:
```diff
@@ -0,0 +1,209 @@
+            }
+            break;
+        case ASTARTE_MAPPING_TYPE_DOUBLE:
+            double dbl = false;
```
same here and in the rest of the function, it is more idiomatic.

sorru94 committed:
Apply suggestions from code review

Signed-off-by: Simone Orru <simone.orru@secomind.com>
